internal/pkg/keeper/client.go

```diff
@@ -0,0 +1,333 @@
+//
+// Copyright (C) 2024 IOTech Ltd
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package keeper
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/edgexfoundry/go-mod-registry/v3/pkg/types"
+)
+
+const (
+	defaultTimeout = 10 * time.Second
+	haltStatus     = "HALT"
+)
+
+type keeperClient struct {
+	config              *types.Config
+	keeperUrl           string
+	serviceKey          string
+	serviceHost         string
+	servicePort         int
+	healthCheckRoute    string
+	healthCheckInterval string
+}
+
+func NewKeeperClient(registryConfig types.Config) (*keeperClient, error) {
+	client := keeperClient{
+		config:     &registryConfig,
+		serviceKey: registryConfig.ServiceKey,
+		keeperUrl:  registryConfig.GetRegistryUrl(),
+	}
+
+	// ServiceHost will be empty when client isn't registering the service
+	if registryConfig.ServiceHost != "" {
+		client.servicePort = registryConfig.ServicePort
+		client.serviceHost = registryConfig.ServiceHost
+		client.healthCheckRoute = registryConfig.CheckRoute
+		client.healthCheckInterval = registryConfig.CheckInterval
+	}
+
+	return &client, nil
+}
+
+func (k *keeperClient) Register() error {
+	if k.serviceKey == "" || k.serviceHost == "" || k.servicePort == 0 ||
+		k.healthCheckRoute == "" || k.healthCheckInterval == "" {
+		return fmt.Errorf("unable to register service with keeper: Service information not set")
+	}
+
+	registrationReq := AddRegistrationRequest{
+		BaseRequest: BaseRequest{
+			Versionable: Versionable{ApiVersion: ApiVersion},
+		},
+		Registration: RegistrationDTO{
+			ServiceId: k.serviceKey,
+			Host:      k.serviceHost,
+			Port:      k.servicePort,
+			HealthCheck: HealthCheck{
+				Interval: k.healthCheckInterval,
+				Path:     k.healthCheckRoute,
+				Type:     "http",
+			},
+		},
+	}
+
+	jsonEncodedData, err := json.Marshal(registrationReq)
+	if err != nil {
+		return fmt.Errorf("failed to encode registration request: %s", err.Error())
+	}
+
+	// check if the service registry exists first
+	resp, err := getRegistryByService(k.config.GetRegistryUrl() + ApiRegistrationByServiceIdRoute + k.serviceKey)
+	if err != nil {
+		return fmt.Errorf("failed to check the %s service registry status: %s", k.serviceKey, err.Error())
+	}
+
+	// call the PUT registry API to update the registry if the service already exists
+	// otherwise, call the POST API to create the registry
+	httpMethod := http.MethodPost
+	if resp.StatusCode == http.StatusOK {
+		httpMethod = http.MethodPut
+	}
+	req, err := http.NewRequest(httpMethod, k.config.GetRegistryUrl()+ApiRegisterRoute, bytes.NewReader(jsonEncodedData))
+	if err != nil {
+		return fmt.Errorf("failed to create register request: %s", err.Error())
+	}
+	req.Header.Set(ContentType, ContentTypeJSON)
+
+	client := http.Client{Timeout: defaultTimeout}
+	resp, err = client.Do(req)
+	if err != nil {
+		return fmt.Errorf("http error: %s", err.Error())
```

Comment:
error message needs more context. Appli similar fix to all error messages in to file that need more context, i.e. the action/api name.
```suggestion
		return fmt.Errorf("failed to register %s: %s",  k.serviceKey, err.Error())
```

Reply:
Done.